src/core/mapping/base_mapper.cc

```diff
@@ -265,108 +213,41 @@ void BaseMapper::slice_auto_task(const MapperContext ctx,
-  if (local_cpus.size() > 0) generate_prime_factor(local_cpus, Processor::LOC_PROC);
-}
-
-const std::vector<int32_t> BaseMapper::get_processor_grid(Legion::Processor::Kind kind,
-                                                          int32_t ndim)
-{
-  auto key    = std::make_pair(kind, ndim);
-  auto finder = proc_grids.find(key);
-  if (finder != proc_grids.end()) return finder->second;
-
-  int32_t num_procs = dispatch(kind, [](auto& procs) { return procs.size(); });
-
-  std::vector<int32_t> grid;
-  auto factor_it = all_factors[kind].begin();
-  grid.resize(ndim, 1);
-
-  while (num_procs > 1) {
-    auto min_it = std::min_element(grid.begin(), grid.end());
-    auto factor = *factor_it++;
-    (*min_it) *= factor;
-    num_procs /= factor;
+  auto lo = key_functor->project_point(sharding_domain.lo(), sharding_domain);
+  auto hi = key_functor->project_point(sharding_domain.hi(), sharding_domain);
+  for (Domain::DomainPointIterator itr(input.domain); itr; itr++) {
+    auto p   = key_functor->project_point(itr.p, sharding_domain);
+    auto idx = linearize(lo, hi, p) % size;
+#ifdef DEBUG_LEGATE
+    assert(idx >= offset);
+#endif
+    output.slices.push_back(TaskSlice(
+      Domain(itr.p, itr.p), avail_procs[idx - offset], false /*recurse*/, false /*stealable*/));
   }
-
-  auto& pitches = proc_grids[key];
-  pitches.resize(ndim, 1);
-  for (int32_t dim = 1; dim < ndim; ++dim) pitches[dim] = pitches[dim - 1] * grid[dim - 1];
-
-  return pitches;
-}
-
-void BaseMapper::slice_manual_task(const MapperContext ctx,
-                                   const LegionTask& task,
-                                   const SliceTaskInput& input,
-                                   SliceTaskOutput& output)
-{
-  output.slices.reserve(input.domain.get_volume());
-
-  auto distribute = [&](auto& procs) {
-    auto ndim       = input.domain.dim;
-    auto& proc_grid = get_processor_grid(task.target_proc.kind(), ndim);
-    for (Domain::DomainPointIterator itr(input.domain); itr; itr++) {
-      int32_t idx = 0;
-      for (int32_t dim = 0; dim < ndim; ++dim) idx += proc_grid[dim] * itr.p[dim];
-      output.slices.push_back(TaskSlice(
-        Domain(itr.p, itr.p), procs[idx % procs.size()], false /*recurse*/, false /*stealable*/));
-    }
-  };
-
-  dispatch(task.target_proc.kind(), distribute);
 }
 
 void BaseMapper::slice_task(const MapperContext ctx,
                             const LegionTask& task,
                             const SliceTaskInput& input,
                             SliceTaskOutput& output)
 {
-  if (task.tag == LEGATE_CORE_MANUAL_PARALLEL_LAUNCH_TAG)
-    slice_manual_task(ctx, task, input, output);
-  else
-    slice_auto_task(ctx, task, input, output);
+  Task legate_task(&task, context, runtime, ctx);
+
+  auto& machine_desc = legate_task.machine_desc();
+  Span<const Processor> avail_procs;
+  uint32_t size;
+  uint32_t offset;
+  machine->dispatch(legate_task.target(), [&](auto target, auto& procs) {
```

Comment:
The `dispatch` pattern here seems unnecessary? You could have:

```
auto [avail_procs, size, offset] = machine_desc.slice(
  legion_task.target(), machine->get_procs(legion_task.target()), 
  machine->total_nodes, machine->local_node
);
```
The `dispatch` method doesn't have a clear name. It also reference *this captures in an asynchronous runtime, which although fine here creates confusion since it's not clear what `dispatch` is doing.

Reply:
The `dispatch` method predates this PR to serve other purposes and I used it just to make quick progress. You will notice that the latest code doesn't have `dispatch` anymore.